Caching HTTP requests tutorial created (#69)
Knotx/knotx-website: src/render/tutorials/caching-http-requests/2_0.html.md

---
title: Caching Proxy for APIs
author: oskarjerzyk
keywords: tutorial
date: 2019-10-21
layout: tutorial
knotxVersions:
    - 2.0.0
---
# Overview

> In this tutorial, we will use Knot.x as API gateway providing additional caching functionality. Knot.x will expose a proxy API to the target one.

Author: fixed

> The caching functionality can be implemented in a custom handler. Then we expose our custom handler in Open API specification. However, this approach requires development effort. But what if we would need to add a circuit breaker mechanism in future.

Author: This change was proposed by @tomaszmichalak therefore I updated PR with it

> The caching functionality can be implemented in a custom handler. Then we expose our custom handler in Open API specification. However, this approach requires development effort. But what if we would need to add a circuit breaker mechanism in future.

Author: fixed

> To get as much as you can from this tutorial, it's highly recommended to get familiar with previous:

Author: fixed

> To get as much as you can from this tutorial, it's highly recommended to get familiar with previous:

Author: fixed

- [Getting started with Knot.x Stack](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/)
- [Getting Started with Docker](http://knotx.io/tutorials/getting-started-with-docker/2_0/)
- [Configurable API gateway](http://knotx.io/tutorials/configurable-api-gateway/2_0/)

## Prerequisites
- JDK 8
- Docker

> - Docker Swarm

Author: fixed

## What you're going to learn:

> - How to wrap an existing (target) API with caching proxy

Author: I agree, it sounds better

> - How to wrap an existing (target) API with caching proxy

Author: fixed

- How to use [behaviours](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#behaviours) and [actions](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#actions)

Download the [Latest Knot.x Starter Kit release](https://github.com/Knotx/knotx-starter-kit/releases) and unzip it.

Project has the following structure:
```
├── docker
|   ├── Dockerfile                // Docker file with image definition.
├── functional                    // Keep here your functional tests. Example implementation included
├── gradle                        // Gradle wrapper and common gradle scripts
├── knotx/conf                    // Knotx configuration which will be copied to docker image
├── modules                       // Sub-modules of your project
│   ├── ...                       // example modules implementation
```

## Actions & behaviours

> The first thing we are going to do is to define our custom task which will properly handle our request. As you probably know,

Author: fixed

> The first thing we are going to do is to define our custom task which will properly handle our request. As you probably know,

Author: fixed

> Knot.x tasks are some kind of directed graphs consisting of actions (if you want to get more familiar with

Author: fixed

> Knot.x tasks are some kind of directed graphs consisting of actions (if you want to get more familiar with

Author: fixed

> configurable integrations please read [this](https://knotx.io/blog/configurable-integrations/) tutorial). Now, let's get it started 

Author: fixed

> configurable integrations please read [this](https://knotx.io/blog/configurable-integrations/) tutorial). Now, let's get it started 

Author: fixed

> and define our task

Author: fixed

```hocon
tasks {
  api-task {
    action = fetch-product-with-cache
    onTransitions {
      _success {
        action = product-to-body
      }
    }
  }
}
```

> As you see above, to achieve our goal we have to create only one task called `api-task` - this is very simple directed graph.

Author: fixed

> As you see above, to achieve our goal we have to create only one task called `api-task` - this is very simple directed graph.

Author: fixed

The action it performs is `fetch-product-with-cache` which we'll create soon. Then we have `onTransitions`, in our case let's 
define only `_success` - the action to be performed is `product-to-body`  which we'll also define.

```hocon
actions {
  fetch-product-with-cache {
    factory = in-memory-cache
    config {
      cache {
        maximumSize = 1000
        ttl = 5000
      }
      cacheKey = "product-{param.id}"
      payloadKey = fetch-product
    }
    doAction = fetch-product
  }

  product-to-body {
    factory = payload-to-body
    config {
      key = fetch-product
    }
  }

  fetch-product {
    factory = http
    config {
      endpointOptions {
        path = /product/id
        domain = webapi
        port = 8080
        allowedRequestHeaders = ["Content-Type"]
      }
    }
  }
}
```

> Above you can see all necessary actions definitions. As mentioned before, we've created `fetch-product-with-cache`. The purpose 

Author: fixed

of this action is to add cache functionality. Now you can see how easily we can define custom behaviours for our actions. 

> We've just assigned to it `in-memory-cache` action factory, moreover this action also wraps other - in this case it's 

Author: fixed

`fetch-product` - it's using simple http factory which will call desired endpoint - in our case `/product/id` (we'll define it later).

> Finally we can focus on `product-to-body` - behaviour of this action is really simple - it's just adding previously fetched data

Author: fixed

> Finally we can focus on `product-to-body` - behaviour of this action is really simple - it's just adding previously fetched data

Author: fixed

to response body.

> Last thing to do here is to combine this two chunks of code into one file, let's call it `fragments-handler.conf ` and put it

Author: fixed

> Last thing to do here is to combine this two chunks of code into one file, let's call it `fragments-handler.conf ` and put it

Author: fixed

inside `knotx/conf/routes/handlers/` directory.

> ## Why we need caching?

Author: fixed

We've just added caching behaviour to action but you may wonder why we did it? Why we're using cache? The general purpose of

> using caching is faster data retrieval - when user call the same endpoint for few times, we don't have to fetch data

Reviewer: ```suggestion
using caching is faster data retrieval - when the user is calling the same endpoint a few times, we don't have to fetch the data
```

Author: fixed

> using caching is faster data retrieval - when user call the same endpoint for few times, we don't have to fetch data

Reviewer 2: ```suggestion
using caching is faster data retrieval - when user calls the same endpoint a few times, we don't have to fetch data
```

Author: fixed

> from server for every request - we can cache it and reuse response (if not changed) in the future - it affects in quicker

Reviewer: ```suggestion
from the server for every request - we can cache and reuse the response in the future (if not changed) - it results in a quicker
```

Author: fixed

> from server for every request - we can cache it and reuse response (if not changed) in the future - it affects in quicker

Reviewer 2: ```suggestion
from the server for every request - we can cache it and reuse the response (if not changed) in the future - it effects in quicker
```

Author: fixed

> response time, what is noticeable by the user. On the other hand, from server side it reduces latency and traffic. Requests

Reviewer: ```suggestion
response time, which improves the overall user experience. Other benefits include reduced latency and traffic. Requests
```

Author: fixed

> response time, what is noticeable by the user. On the other hand, from server side it reduces latency and traffic. Requests

Reviewer 2: ```suggestion
response time that's noticeable by the user. On the other hand, at the server side it reduces latency and traffic. Requests
```

Author: fixed

> are satisfied in a shorter time with reused representations.

Reviewer: ```suggestion
are satisfied in a shorter time with cached values.
```

Author: fixed

> You may also wonder if we can replace this cache with another one, for example with Redis - yes you can do it very simple!

Reviewer: provide a link to Redis and rephrase the `You may also wonder` part

Author: fixed

> You may also wonder if we can replace this cache with another one, for example with Redis - yes you can do it very simple!

Reviewer 2: ```suggestion
You may also wonder if we can replace this cache with another one, for example with Redis. The answer is yes, you can do it very simply!
```

Author: fixed

> The only thing needed here is to create proper action factory which will handle desired cache and properly link it to action

Reviewer 2: ```suggestion
The only thing needed here is to create appropriate custom action factory which will handle caching in the desired way and then properly link it to an action
```

Author: fixed

> in config file (the same way is we did above).

Reviewer 2: ```suggestion
in the config file (the same way is we did above).
```

Author: fixed

## Operations configuration

> Now we have to configure `knotx/conf/routes/operations.conf`. In our config we should have define two operations:

Reviewer: ```suggestion
Now we have to configure `knotx/conf/routes/operations.conf` and define two operations:
```

Author: fixed

> Now we have to configure `knotx/conf/routes/operations.conf`. In our config we should have define two operations:

Reviewer 2: ```suggestion
Now we have to configure `knotx/conf/routes/operations.conf`. In our config we should have two operations defined:
```

Author: fixed

- `healthcheck-operation`
- `product-api-caching-proxy-operation`

> This operations are connected with `openapi.yaml` which we'll create next. This file specifies which handlers should be invoked 

Reviewer: ```suggestion
These operations are connected with `openapi.yaml` which we'll create next. This file specifies which handlers should be invoked 
```

Author: fixed

> to satisfy request. Now we'll focus on operation called `product-api-caching-proxy-operation` (it will be connected to `openapi.yaml`).

Reviewer: ```suggestion
to process the request. Now we'll focus on operation called `product-api-caching-proxy-operation` (it will be connected to `openapi.yaml`).
```

Author: fixed

> to satisfy request. Now we'll focus on operation called `product-api-caching-proxy-operation` (it will be connected to `openapi.yaml`).

Reviewer 2: ```suggestion
to satisfy requests. Now we'll focus on the operation called `product-api-caching-proxy-operation` (it will be connected to `openapi.yaml`).
```

Reviewer 2: Also, I think saying that "operations are connected with openapi.yaml" may be confusing

Author: fixed

Below you can see file definition:

```hocon
routingOperations = ${routingOperations} [
  {
    operationId = product-api-caching-proxy-operation
    handlers = ${config.server.handlers.common.request} [
      {
        name = singleFragmentSupplier
        config = {
          type = json
          configuration {
            data-knotx-task = api-task
          }
        }
      },
      {
        name = fragmentsHandler
        config = {include required(classpath("routes/handlers/fragments-handler.conf"))}
      },

      {
        name = fragmentsAssembler
      }
    ] ${config.server.handlers.common.response}
  }
  {
    operationId = healthcheck-operation
    handlers = [
      {
        name = healthcheck
      }
    ]
  }
]
```

> Regarding mentioned operation, we can see, that we're using three different handlers to perform it.

Reviewer: ```suggestion
As shown above, we're using three different handlers to perform it.
```

Author: fixed

> In our case, first handler to be called is `singleFragmentSupplier`. It's used to convert incoming http request

Reviewer: ```suggestion
The first handler to be called is `singleFragmentSupplier`. It's used to convert incoming http request
```

Author: fixed

into `knotx-fragment` and assign `data-knotx-task` to it. Next to be called is `fragmentsHandler`, here we have to attach
additional configs with tasks specification. The last one is `fragmentsAssembler`

> Purpose of using it, is to combine all fragments into final response. And this is all about operations.

Reviewer: ```suggestion
It combines all fragments into a final response.
```

Author: fixed

> Purpose of using it, is to combine all fragments into final response. And this is all about operations.

Reviewer 2: ```suggestion
The purpose of using it, is to combine all fragments into the final response. And this is all when it comes to operations.
```

Author: fixed

## OpenAPI
Finally we can create `./knotx/conf/openapi.yaml`, it contains project-specific API definitions using the [OpenAPI 3.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md) syntax. 
Let's fill it out with:
```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: API gateway caching example
  description: API gateway caching example

servers:
  - url: https://{domain}:{port}
    description: The local API server
    variables:
      domain:
        default: localhost
        description: api domain
      port:
        enum:
          - '8092'
        default: '8092'

paths:
  /healthcheck:
    get:
      operationId: healthcheck-operation
      responses:
        default:
          description: example vert.x healthcheck

  /product/id:
    get:
      operationId: product-api-caching-proxy-operation
      responses:
        default:
          description: Target API (Product API) caching proxy.
```

> This is whole structure of this file required in our case. In more detail it was explained in [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/) 

Reviewer: ```suggestion
This is whole structure of this file required in our case. It was explained in more details in [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/) 
```

Author: fixed

> This is whole structure of this file required in our case. In more detail it was explained in [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/) 

Reviewer 2: ```suggestion
This is the whole structure of this file required in our case. It is explained in more details in [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/) 
```

Author: fixed

> The most important for us here is `/product/id` endpoint definition. As you can see we are connecting this endpoint with

Reviewer: ```suggestion
The most important part is `/product/id` endpoint definition. As you can see we are connecting this endpoint with
```

Author: fixed

> The most important for us here is `/product/id` endpoint definition. As you can see we are connecting this endpoint with

Reviewer 2: ```suggestion
The most important thing for us here is `/product/id` endpoint definition. As you can see we are connecting this endpoint with
```

Author: fixed

> `product-api-caching-proxy-operation` which we've defined before. Moreover here we've provided some other specification 

Reviewer: Delete the `Moreover ...` sentence

Author: fixed

> but it was described in other tutorials, so if you don't feel familiar with it, please check them out.

Reviewer 2: ```suggestion
that is described in other tutorials. Please check the tutorials section if you feel you need more explanation.
```

Author: fixed

> We have all Knot.x specific files created, now we are gonna focus on project setup and deployment.  

Reviewer 2: ```suggestion
We have all Knot.x specific files created, now we are going to focus on project setup and deployment.  
```
"gonna" is a slang

Author: fixed

## Project configuration

> The first thing we need to do is to specify configuration files. Let's start in the project level directory and simply create `api-cache.yml`

Reviewer: A few words on what file we're creating here (Docker swarm config) would be nice.

Author: fixed

Let's specify the following configuration:
``` yml
version: '3.7'

networks:
  knotnet:

services:

  webapi:
    image: rodolpheche/wiremock
    volumes:
      - "./common-services/webapi:/home/wiremock"
    ports:
      - "3000:8080"
    networks:
      - knotnet

  knotx:
    image: knotx-example/api-cache:latest
    command: ["knotx", "run-knotx"]
    ports:
      - "8092:8092"
      - "18092:18092"
    networks:
      - knotnet
```

First and foremost it specifies Docker virtual network in which our services will be deployed. Then we have some service configs.
The first one is `webapi`
- `image` - docker image is provided here, 
- `volumes` - directories to which Docker will have access, in our case it's directory where all Wiremock configs are stored,

> - `ports` - on the left side is provided port accessible from `localhost`, and on the right side we have port number available in virtual environment 

Reviewer 2: ```suggestion
- `ports` - port accessible from `localhost` is specified on the left side, and a port available in the virtual environment is specified on the right
```

Author: fixed

> - `networks` - specifies in which docker network, service will be deployed.

Reviewer 2: ```suggestion
- `networks` - specifies in which docker network the service will be deployed.
```

Author: fixed

> Then we have `knotx` service specification, it's very similar to previous one, so only one property is worth to explain 

Reviewer 2: ```suggestion
Then we have `knotx` service specification. It's very similar to the previous one, so only one property is worth explaning
```

> - `command` - this property is used to point Knot.x starting commands.

Reviewer 2: ```suggestion
- `command` - this property is used to specify Knot.x starting commands.
```

> Next step in our configuration is to properly set up `settings.gradle.kts` (also at the project level directory) 

Reviewer: It's already in Starter Kit.

Author: paragraph changed a bit, I agree that's in starterkit but we have to mention it because user has to specify rootProject.name there

> Next step in our configuration is to properly set up `settings.gradle.kts` (also at the project level directory) 

Reviewer 2: Overall comment to this part of the tutorial. I feel like there to much details here. Aren't those things explained in Getting Started With Docker?

The same if it comes to `./docker/Dockerfile`, version provided by default is absolutely enough for our purposes.

## Wiremock 
Wiremock is simply used for mocking server, we don't have to create real REST API, for our training purposes it's enough.
To do it let's create following directories on the project level directory:
- `common-services/webapi/__files/`
- `common-services/webapi/mappings/`

Into first directory please paste `product.json` file with following body:
```json
{
  "id": 21762532,
  "url": "http://knotx.io",
  "label": "Product"
}
```

And into second directory please paste `product.json` file with following body:
```json
{
  "request": {
    "method": "GET",
    "url": "/product/id"
  },
  "response": {
    "status": 200,
    "fixedDelayMilliseconds": 100,
    "bodyFileName": "product.json"
  }
}
```

Make sure that in `api-cache.yml` property `volume` properly points to `webapi` directory. And that's all about Wiremock.

## Build and run
From project level directory execute following commands which will build project and run Docker Swarm:
```shell script
$ gradlew clean build
$ docker swarm init
$ docker stack deploy -c api-cache.yml api-cache
```

> After successful execution you should be able to see http response at [localhost:8092/product/id](localhost:8092/product/id)

Reviewer 2: Some summary of what we did and what we can take from this would be nice. Like... moral of the story :D

Author: done

We've just assigned `in-memory-cache` action factory to it Moreover this action also wraps another one - in this case it's 
`fetch-product`. It's using simple http factory which will call desired endpoint, in our case `/product/id` (we'll define it later).
Finally we can focus on `product-to-body` - it's just adding previously fetched data to response body.

## Why do we need caching?
We've just added caching behaviour to action but you may wonder what are the benefits? Why we're using cache? The general purpose of
using caching is faster data retrieval - when the user is calling the same endpoint a few times, we don't have to fetch the data
from the server for every request - we can cache and reuse the response in the future (if not changed) - it effects in a quicker
response time, which improves the overall user experience. Other benefits include reduced latency and traffic. Requests
are satisfied in a shorter time with cached values.

You may also wonder if we can replace this cache with another one, for example with [Redis](https://redis.io/). The answer is yes, you can do it very simply!

> The only thing needed here is to create appropriate custom action factory which will handle caching in the desired way and then properly link it to an action

Reviewer 2: I thought we wanted to provide an example of redia implementation with Knot.x?

Author: probably we'll provide it in the future

## Wiremock 
Wiremock is simply used for mocking server, we don't have to create real REST API, for our training purposes it's enough.
To do it let's create following directories on the project level directory:
- `common-services/webapi/__files/`
- `common-services/webapi/mappings/`

Into first directory please paste `product.json` file with following body:
```json
{
  "id": 21762532,
  "url": "http://knotx.io",
  "label": "Product"
}
```

And into second directory please paste `product.json` file with following body:
```json
{
  "request": {
    "method": "GET",
    "url": "/product/id"
  },
  "response": {
    "status": 200,
    "fixedDelayMilliseconds": 100,
    "bodyFileName": "product.json"
  }
}
```

Make sure that in `api-cache.yml` property `volume` properly points to `webapi` directory. And that's all about Wiremock.

## Build and run
From project level directory execute following commands which will build project and run Docker Swarm:
```shell script
$ gradlew clean build
$ docker swarm init
$ docker stack deploy -c api-cache.yml api-cache
```
After successful execution you should be able to see http response at [localhost:8092/product/id](localhost:8092/product/id)

> ## Summary of what we actually did

Reviewer: ```suggestion
## Summary
```

Author: fixed

> To sum it up in this tutorial we've configured in-memory cache and show how easily we can replace it with another one.

Reviewer: ```suggestion
In this tutorial we've configured in-memory cache and we've shown how easily we can replace it with another one.
```

Author: fixed

> What's important we did it without writing any line of code, but only using Knot.x configs. We've focused on creating it

Reviewer: ```suggestion
The important thing is: we did it without writing any line of code, using only Knot.x configs. We've focused on creating it
```

Author: fixed

> in declarative way instead of implementing cache on our own. Hopefully, now you'll know how easily you can extend other 

Reviewer: `Hopefully ... ` sounds quite bad, it suggests that the reader might not understand the tutorial

Author: fixed

Wiremock is simply used for mocking server, we don't have to create real REST API, for our training purposes it's enough.
To do it let's create following directories on the project level directory:
- `common-services/webapi/__files/`
- `common-services/webapi/mappings/`

Into first directory please paste `product.json` file with following body:
```json
{
  "id": 21762532,
  "url": "http://knotx.io",
  "label": "Product"
}
```

And into second directory please paste `product.json` file with following body:
```json
{
  "request": {
    "method": "GET",
    "url": "/product/id"
  },
  "response": {
    "status": 200,
    "fixedDelayMilliseconds": 100,
    "bodyFileName": "product.json"
  }
}
```

Make sure that in `api-cache.yml` property `volume` properly points to `webapi` directory. And that's all about Wiremock.

## Build and run
From project level directory execute following commands which will build project and run Docker Swarm:
```shell script
$ gradlew clean build
$ docker swarm init
$ docker stack deploy -c api-cache.yml api-cache
```
After successful execution you should be able to see http response at [localhost:8092/product/id](localhost:8092/product/id)

## Summary

> In this tutorial we've configured in-memory cache and we've shown how easily we can replace it with another one.

Reviewer 3: What do you think about this:
> In this tutorial, we used Knot.x as a simple API gateway providing very flexible API caching functionality. With [Configurable Integrations](https://knotx.io/blog/configurable-integrations/) we
achieved it without writing any line of code, using only Knot.x built-in features. With [Actions](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#actions) and [Behaviours](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#behaviours), we've focused on creating it in a declarative way instead of implementing caching functionality on our own. Now you'll understand how quickly you can "decorate" your existing API with cross-cutting functionalities such as caching or circuit breaker.

Author: looks better, applied

The important thing is: we did it without writing any line of code, using only Knot.x configs. We've focused on creating it
in declarative way instead of implementing cache on our own. Now you'll know how easily you can extend other 
actions with cache functionality and understand what benefits it brings.

> You can find the complete code of this tutorial [in our github repository](https://github.com/Knotx/knotx-example-project/tree/master/api-gateway/)

Reviewer 3: Can we point to api-cache ?

Author: fixed

---
title: Caching Proxy for APIs
author: oskarjerzyk
keywords: tutorial
date: 2019-10-21
layout: tutorial
knotxVersions:
    - 2.0.0
---
# Overview

> In this tutorial, we will use Knot.x as API gateway in order to show the usage of caching functionality. Knot.x will act as a proxy API to some external API and will cache it's responses.

Author: fixed

> The caching functionality can be implemented in a custom handler. We would then expose our custom handler in the OpenAPI file. However, this approach would require development effort. And what if we would need to add a circuit breaker mechanism in the future?

Author: fixed

> The solution is simple - tasks and dedicated `in-memory cache` action. We'll do it in this tutorial. 

Author: fixed

- [Getting Started with Docker](http://knotx.io/tutorials/getting-started-with-docker/2_0/)
- [Configurable API gateway](http://knotx.io/tutorials/configurable-api-gateway/2_0/)

## Prerequisites
- JDK 8
- Docker

## What you're going to learn:
- How to wrap an existing (external) API with a caching proxy
- How to use [behaviours](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#behaviours) and [actions](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#actions)

Download the [Latest Knot.x Starter Kit release](https://github.com/Knotx/knotx-starter-kit/releases) and unzip it.

Project has the following structure:
```
├── docker
|   ├── Dockerfile                // Docker file with image definition.
├── functional                    // Keep here your functional tests. Example implementation included
├── gradle                        // Gradle wrapper and common gradle scripts
├── knotx/conf                    // Knotx configuration which will be copied to docker image
├── modules                       // Sub-modules of your project
│   ├── ...                       // example modules implementation
```

## Actions & behaviours

> The first thing we are going to do, is define our custom task which will properly handle our request. As explained in the previous tutorials,

Author: sounds better, fixed

    action = fetch-product-with-cache
    onTransitions {
      _success {
        action = product-to-body
      }
    }
  }
}

actions {
  # Target API caching proxy
  fetch-product-with-cache {
    # https://github.com/Knotx/knotx-fragments/blob/2.0.0/handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java
    factory = in-memory-cache
    config {
      cache {
        maximumSize = 1000
        ttl = 5000
      }
      cacheKey = "product-{param.id}"
      payloadKey = fetch-product
    }
    doAction = fetch-product
  }

  # Target API
  fetch-product {
    # https://github.com/Knotx/knotx-data-bridge/blob/2.0.0/http/action/src/main/java/io/knotx/databridge/http/action/HttpActionFactory.java
    factory = http
    config {
      endpointOptions {
        path = /product/id
        domain = webapi
        port = 8080
        allowedRequestHeaders = ["Content-Type"]
      }
    }

  }

  product-to-body {
    # https://github.com/Knotx/knotx-fragments/blob/2.0.0/handler/core/src/main/java/io/knotx/fragments/handler/action/PayloadToBodyActionFactory.java
    factory = payload-to-body
    config {
      key = fetch-product
    }
  }
}
```

> As you can see above, to achieve our goal we have to create only one task called `get-product-task` - this is very simple directed graph.

Reviewer 3: Maybe:
>As you can see above, we declare the `get-product-task` task that points to the `fetch-product-with-cache` action node. The action declares one `success` transition (outgoing graph edge) that points to the `product-to-body` action node.

Reviewer 3: You specify actions in the same paragraph, you do not have to mention that `you would describe them later`.

Author: fixed

The action it performs is `fetch-product-with-cache` which we'll create soon. Then we have `onTransitions`, in our case let's 
define only `_success` - the action to be performed is `product-to-body`  which we'll also define.

> Below task definition you can see all necessary action's definitions. As mentioned before, we've created `fetch-product-with-cache`. The purpose 

Author: fixed

> of this action is to add cache functionality. Now you can see how easily we can define custom behaviours for our actions. 

Author: I agree, it looks better, code replaced with your changes

We've just assigned `in-memory-cache` action factory to it Moreover this action also wraps another one - in this case it's 
`fetch-product`. It's using simple http factory which will call desired endpoint, in our case `/product/id` (we'll define it later).
Finally we can focus on `product-to-body` - it's just adding previously fetched data to response body.

## Why do we need caching?

> We've just added caching behaviour to action but you may wonder what are the benefits? Why we're using cache? The general purpose of

Author: fixed

using caching is faster data retrieval - when the user is calling the same endpoint a few times, we don't have to fetch the data

> from the server for every request - we can cache and reuse the response in the future (if not changed) - it effects in a quicker

Author: fixed

> response time, which improves the overall user experience. Other benefits include reduced latency and traffic. Requests

Author: fixed

are satisfied in a shorter time with cached values.

You may also wonder if we can replace this cache with another one, for example with [Redis](https://redis.io/). The answer is yes, you can do it very simply!
The only thing needed here is to create appropriate custom action factory which will handle caching in the desired way and then properly link it to an action
in the config file (the same way is we did above).

## Operations configuration
Now we have to configure `knotx/conf/routes/operations.conf` and define two operations:
- `healthcheck-operation`
- `product-api-caching-proxy-operation`

> These operations are connected with `openapi.yaml` which we'll create next. This file specifies which handlers should be invoked 

Reviewer 3: ```suggestion
These operations are configured in the `openapi.yaml` specification which we'll do next. This file specifies which handlers should be invoked 
```

Author: fixed

> to process the requests. Now we'll focus on the operation called `product-api-caching-proxy-operation` (it will be connected to `openapi.yaml`).

Reviewer 3: ```suggestion
to process the requests. Let's specify the `product-api-caching-proxy-operation` operation first:
```

Author: fixed

> Below you can see file definition:

Reviewer 3: This can be removed.

Author: fixed

```hocon
routingOperations = ${routingOperations} [
  {
    operationId = product-api-caching-proxy-operation
    handlers = ${config.server.handlers.common.request} [
      {
        name = singleFragmentSupplier
        config = {
          type = json
          configuration {
            data-knotx-task = get-product-task
          }
        }
      },
      {
        name = fragmentsHandler
        config = { include required(classpath("routes/handlers/fragments-handler.conf")) }
      },

      {
        name = fragmentsAssembler
      }
    ] ${config.server.handlers.common.response}
  }
  {
    operationId = healthcheck-operation
    handlers = [
      {
        name = healthcheck
      }
    ]
  }
]
```

> As shown above, we're using three different handlers to perform it.

Reviewer 3: >The operation consists of three handlers:
> - `singleFragmentSupplier` that converts incoming HTTP request to [Fragment](https://github.com/Knotx/knotx-fragments/tree/2.0.0/api#knotx-fragment-api) and assign a task to it
> - `fragmentsHandler` handler that reads task configuration and evaluates graph logic
> - `fragmentsAssembler` handler that rewrite fragment's body into the final HTTP response

Author: fixed

Let's specify the following configuration:
``` yml
version: '3.7'

networks:
  knotnet:

services:

  webapi:
    image: rodolpheche/wiremock
    volumes:
      - "./common-services/webapi:/home/wiremock"
    ports:
      - "3000:8080"
    networks:
      - knotnet

  knotx:
    image: knotx-example/api-cache:latest
    command: ["knotx", "run-knotx"]
    ports:
      - "8092:8092"
      - "18092:18092"
    networks:
      - knotnet
```

First and foremost it specifies Docker virtual network in which our services will be deployed. Then we have some service configs.
The first one is `webapi`
- `image` - docker image is provided here, 
- `volumes` - directories to which Docker will have access, in our case it's directory where all Wiremock configs are stored,
- `ports` - port accessible from `localhost` is specified on the left side, and a port available in the virtual environment is specified on the right 
- `networks` - specifies in which docker network the service will be deployed.

Then we have `knotx` service specification. It's very similar to the previous one, so only one property is worth explaining. 
- `command` - this property is used to specify Knot.x starting commands.

> Next step in our configuration is to set `rootProject.name` in `settings.gradle.kts` (also at the project level directory),

Reviewer 3: I am not sure if this rename is required for this tutorial. The "api-cache" name is not connected with docker image name.

Author: it will of course work without it, but I think that we can provide proper name for it as it is

also make sure that `healh-check` is included 
```gradle
rootProject.name = "api-cache"

include("health-check")

project(":health-check").projectDir = file("modules/health-check")
``` 

> Next thing to do is to configure `gradle.properties` at project level directory. Let's fill it out with

Reviewer 3: Let's define the custom Docker image name for Knot.x instance. We can specify it by the `docker.image.name` property in the `./gradle.properties` file:

Author: fixed

Author: fixed

```gradle
version=2.0.0-SNAPSHOT
knotx.version=2.0.0
knotx.conf=knotx
docker.image.name=knotx-example/api-cache
```

> Very important step here is to set `docker.image.name`, which is necessary to run proper docker image.

Reviewer 3: To remove.

Author: fixed

Author: fixed

> `build.gradle.kts` file could stay as it was provided with starterkit, we don't have to apply any modifications here.

Reviewer 3: Remove.

Author: fixed

> The same if it comes to `./docker/Dockerfile`, version provided by default is absolutely enough for our purposes.

Reviewer 3: Remove.

Author: fixed

## Wiremock 

> Wiremock is simply used for mocking server, we don't have to create real REST API, for our training purposes it's enough.

Reviewer 3: ```suggestion
We use Wiremock to mock our target API behaviour. We can easily specify the response body, code and delay. To do it let's create following directories:
```

Author: fixed

To do it let's create following directories on the project level directory:

> - `common-services/webapi/__files/`

Reviewer 3: ```suggestion
- `./common-services/webapi/__files/`
```

Author: fixed

> - `common-services/webapi/mappings/`

Reviewer 3: ```suggestion
- `./common-services/webapi/mappings/`
```

Author: fixed

> Into first directory please paste `product.json` file with following body:

Reviewer 3: ```suggestion
In the `./common-services/webapi/__files/` directory please create the `product.json` file with following body:
```

Author: fixed

```json
{
  "id": 21762532,
  "url": "http://knotx.io",
  "label": "Product"
}
```

> And into second directory please paste `product.json` file with following body:

Reviewer 3: ```suggestion
In the `./common-services/webapi/mappings/` directory please create the `product.json` file with following body:
```

Author: fixed

```json
{
  "request": {
    "method": "GET",
    "url": "/product/id"
  },
  "response": {
    "status": 200,
    "fixedDelayMilliseconds": 100,
    "bodyFileName": "product.json"
  }
}
```

> Make sure that in `api-cache.yml` property `volume` properly points to `webapi` directory. And that's all about Wiremock.

Reviewer 3: ```suggestion
Make sure that the `volume` property from `./api-cache.yml` points to the `./common-services/webapi` directory. And that's all about Wiremock.
```

Author: fixed

---
title: Caching Proxy for APIs
author: oskarjerzyk
keywords: tutorial
date: 2019-10-21
layout: tutorial
knotxVersions:
    - 2.0.0
---
# Overview
In this tutorial, we will use Knot.x as the API gateway to show the usage of caching functionality. Knot.x will act as a proxy API to some external API and will cache its responses.
The caching functionality can be implemented in a custom handler. We would implement API invocation logic and code caching logic using, for example, Guava cache. 
However, this approach implies the need to implement both the API call and the cache mechanism. Moreover, what happens if we would have to implement a circuit breaker pattern for proxied API in the future? How will it affect our existing integration logic?

> In this tutorial, we would make use of Configurable Integrations. Thanks to this would encapsulate our API invocation with an [HTTP Action](https://github.com/Knotx/knotx-data-bridge/tree/2.0.0/http#http-action), then wrap it with [In-memory Cache Action](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#in-memory-cache-action) and finally specify as a task. Simply? So let's configure it. 

Author: 'will' will be better?

> In this tutorial, we would make use of Configurable Integrations. Thanks to this would encapsulate our API invocation with an [HTTP Action](https://github.com/Knotx/knotx-data-bridge/tree/2.0.0/http#http-action), then wrap it with [In-memory Cache Action](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#in-memory-cache-action) and finally specify as a task. Simply? So let's configure it. 

Author: fixed

> To fully understand this tutorial, it's highly recommended to get familiar with previous ones:

Author: fixed

Finally, we define the `product-to-body` action to rewrite the previously fetched data from the fragment's payload to the response body.

## Why do we need caching?
We've just declared caching behaviour to the API invocation but you may wonder what are the benefits? Why we're using cache? The general purpose of
using caching is faster data retrieval - when the user is calling the same endpoint a few times, we don't have to fetch the data
from the server for every request - we can cache and reuse the response for incoming requests (until cached values are valid) - it allows quicker
response times, which improve the overall user experience. Other benefits include reduced latency and traffic. Requests
are satisfied in a shorter time with cached values.

You may also wonder if we can replace this cache with another one, for example with [Redis](https://redis.io/). The answer is yes, you can do it very simply!
The only thing needed here is to create appropriate custom action factory which will handle caching in the desired way and then properly link it to an action

> in the config file (the same way is we did above).

Author: fixed

Below you can see file definition:

```hocon
routingOperations = ${routingOperations} [
  {
    operationId = product-api-caching-proxy-operation
    handlers = ${config.server.handlers.common.request} [
      {
        name = singleFragmentSupplier
        config = {
          type = json
          configuration {
            data-knotx-task = get-product-task
          }
        }
      },
      {
        name = fragmentsHandler
        config = { include required(classpath("routes/handlers/fragments-handler.conf")) }
      },

      {
        name = fragmentsAssembler
      }
    ] ${config.server.handlers.common.response}
  }
  {
    operationId = healthcheck-operation
    handlers = [
      {
        name = healthcheck
      }
    ]
  }
]
```
As shown above, we're using three different handlers to perform it.
The first handler to be called is `singleFragmentSupplier`. It's used to convert incoming http request
into `knotx-fragment` and assign `data-knotx-task` to it. Next to be called is `fragmentsHandler`, here we have to attach
additional configs with tasks specification. The last one is `fragmentsAssembler`
It combines all fragments into a final response.

## OpenAPI

> Finally we can create `./knotx/conf/openapi.yaml`, it contains project-specific API definitions using the [OpenAPI 3.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md) syntax. 

Reviewer 3: > Finally, we need to declare the exposed the new API containing caching functionality, its path and HTTP method. We add it in the `./knotx/conf/openapi.yaml` [OpenAPI 3.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md):

Author: fixed

Let's fill it out with:
```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: API gateway caching example
  description: API gateway caching example

servers:
  - url: https://{domain}:{port}
    description: The local API server
    variables:
      domain:
        default: localhost
        description: api domain
      port:
        enum:
          - '8092'
        default: '8092'

paths:
  /healthcheck:
    get:
      operationId: healthcheck-operation
      responses:
        default:
          description: example vert.x healthcheck

  /product/id:
    get:
      operationId: product-api-caching-proxy-operation
      responses:
        default:
          description: External API (Product API) caching proxy.
```

> This is the whole structure of this file required in our case. It was explained in more details in [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/)  

Reviewer 3: I would simplify it:
> The API is exposed by the `product/id` path and GET HTTP method. It delegates its processing logic to the `product-api-caching-proxy-operation` operation.

Author: fixed

The most important part here is `/product/id` endpoint definition. As you can see we are connecting this endpoint with
`product-api-caching-proxy-operation` which we've defined before.

We have all Knot.x specific files created, now we are going to focus on project setup and deployment.  

> ## Project configuration

Reviewer 3: ```suggestion
## Environment configuration
```

Author: fixed

> The first thing we need to do is to specify configuration files. Let's start in the project level directory and simply create `api-cache.yml`

Author: fixed

    command: ["knotx", "run-knotx"]
    ports:
      - "8092:8092"
      - "18092:18092"
    networks:
      - knotnet
```

First and foremost it specifies Docker virtual network in which our services will be deployed. Then we have some service configs.
The first one is `webapi`
- `image` - docker image is provided here, 
- `volumes` - directories to which Docker will have access, in our case it's directory where all Wiremock configs are stored,
- `ports` - port accessible from `localhost` is specified on the left side, and a port available in the virtual environment is specified on the right 
- `networks` - specifies in which docker network the service will be deployed.

Then we have `knotx` service specification. It's very similar to the previous one, so only one property is worth explaining. 
- `command` - this property is used to specify Knot.x starting commands.

Next step in our configuration is to set `rootProject.name` in `settings.gradle.kts` (also at the project level directory),
also make sure that `healh-check` is included 
```gradle
rootProject.name = "api-cache"

include("health-check")

project(":health-check").projectDir = file("modules/health-check")
``` 

> Next thing to do is to configure `gradle.properties` at project level directory. Let's fill it with

Reviewer 2: ```suggestion
Next thing to do is to configure `gradle.properties` at project level directory. Let's fill it out with
```

Author: fixed

> Next thing to do is to configure `gradle.properties` at project level directory. Let's fill it with

Reviewer 3: ```suggestion
Next we update the `./gradle.properties` file:
```

Author: fixed

Author: fixed

---
title: Caching HTTP requests

> author: oskarjerzyk

Reviewer: I think you also have to provide file in `team/contributors`

Author: fixed

keywords: tutorial
date: 2019-10-21
layout: tutorial
knotxVersions:
    - 2.0.0
---
# Overview
In this tutorial we will add ability to cache http request to project based on [Knot.x Starter Kit](https://github.com/Knotx/knotx-starter-kit) 
template.

We will use existing `ActionFactory` from [knotx-fragments](https://github.com/Knotx/knotx-fragments) to achieve our goal
which will be implementing custom endpoint with cache ability.

## Prerequisites
- JDK 8
- Docker

## What you're going to learn:
- How to add cache ability to API
- How to set up Wiremock (we don't need to implement real API, mock API will be enough for our training purposes)

> - How to deploy project to Docker

Reviewer: ```suggestion
- How to deploy project to Docker Swarm
```

Author: fixed

Download the [Latest Knot.x Starter Kit release](https://github.com/Knotx/knotx-starter-kit/releases) and unzip it.

Project has the following structure:
```
├── docker
|   ├── Dockerfile                // Docker file with image definition.
├── functional                    // Keep here your functional tests. Example implementation included
├── gradle                        // Gradle wrapper and common gradle scripts
├── knotx/conf                    // Knotx configuration which will be copied to docker image
├── modules                       // Sub-modules of your project
│   ├── ...                       // example modules implementation
```

## Project configuration
The first thing we need to do is to specify configuration files. Let's start in the project level directory and simply create `api-cache.yml`

> The file should contains

Reviewer: ```suggestion
Let's specify the following configuration:
```

Author: fixed

``` yml
version: '3.7'

networks:
  knotnet:

services:

  webapi:
    image: rodolpheche/wiremock
    volumes:
      - "./common-services/webapi:/home/wiremock"
    ports:
      - "3000:8080"
    networks:
      - knotnet

  knotx:
    image: knotx-example/api-cache:latest
    command: ["knotx", "run-knotx"]
    ports:
      - "8092:8092"
      - "18092:18092"
    networks:
      - knotnet
```

> First and foremost it specifies docker virtual network in which our services will be deployed. Then we have some service configs.

Reviewer: ```suggestion
First and foremost it specifies Docker virtual network in which our services will be deployed. Then we have some service configs.
```

Author: fixed

> The first one is `webapi`, `image` - docker image is provided here, `volumes` - directories to which Docker will have access, 

Reviewer: maybe instead of iterating over commas, a punctuation would be more readable here

Author: fixed

in our case it's directory where all Wiremock configs are stored,
`ports` - on the left side is provided port accessible from `localhost`, and on the right side we have port number available in 
virtual environment. The last param `networks` specifies in which docker network service will be deployed.
Then we have `knotx` service specification, it's very similar to previous one. `command` property is used to point Knot.x starting commands.

> Next step in our configuration is to properly set up `settings.gradle.kts` (also at the project level directory) 

Reviewer: I don't think that's really necessary. The file comes OOTB from Starter Kit and renaming the project is quite trivial

```gradle
rootProject.name = "api-cache"

include("health-check")

project(":health-check").projectDir = file("modules/health-check")
``` 

Here we have to specify project name, and packages to be included in this project. 

Next thing to do is to configure `gradle.properties` at project level directory. Let's fill it with

> ```gradle

Reviewer: maybe let's just limit the config change to renaming the Docker image, that way it will be interoperable with future Knot.x versions

version=2.0.0-SNAPSHOT
knotx.version=2.0.0
knotx.conf=knotx
docker.image.name=knotx-example/api-cache
```
Very important step here is to set `docker.image.name`, which is necessary to run proper docker image.

> `build.gradle.kts` file could stay as it was provided with starterkit, we don't have to apply any modifications here.

Reviewer: I don't think that's not necessary

The same if it comes to `./docker/Dockerfile`, version provided by default is absolutely enough for our purposes.

## Modules

> Now we can focus on modules - with starterkit some are provided. Very important is `health-check`. Role of this module is quite 

Reviewer: `healthcheck` is already described in _Getting started with Docker_ tutorial, you can simply provide a link

self explanatory, we were pointing it in one of the previous files to be included in project image.

> In our case we don't need to prepare custom modules with custom business logic, instead of it we'll create Knot.x configs, 

Reviewer: I don't quite get that sentence o.O

which will set up cache ability in our project.

## Knot.x configs
The first thing to do here is to fill `./knotx/conf/openapi.yaml`. it contains project-specific API definitions using the [OpenAPI 3.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md) syntax. 

> Let's fill it out with:

Reviewer: Could you rephrase it? I don't like the `fill it out` in regard of a file :D

> ```yaml

Reviewer: to make it shorter, I think you could just include the fragment with `paths` or even `product/id`

servers:
  - url: https://{domain}:{port}
    description: The local API server
    variables:
      domain:
        default: localhost
        description: api domain
      port:
        enum:
          - '8092'
        default: '8092'

paths:
  /healthcheck:
    get:
      operationId: healthcheck-operation
      responses:
        default:
          description: example vert.x healthcheck

  /product/id:
    get:
      operationId: example-cacheable-api-operation
      responses:
        default:
          description: Example product with cache operation
```
This is whole structure of this file required in our case. In more detail it was explained in [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/) 
The most important part for us is `paths`. There we have endpoints specification. `/product/id` is the endpoint which we'll call
and cache response from it. `operationId` field specifies which operation should be executed when endpoint is called, 
in our case it will be `example-cacheable-api-operation`. Now we can focus on it and configure everything, what's necessary to achieve proper behaviour
of our endpoint. 

> #### Logger configuration

Reviewer: Logback comes out of the box from Starter Kit, remove this chapter

There is one additional step to improve our work with Docker, we can setup logger properly to get better stack trace when
an error occurs. To do it just go to `knotx/conf/logback.xml` and replace default `configuration` tag with:

```xml
<configuration>
  <property name="LOG_PATH" value="/var/log/knotx"/>

  <include resource="io/knotx/logging/logback/defaults.xml"/>
  <include resource="io/knotx/logging/logback/file-appender.xml"/>
  <include resource="io/knotx/logging/logback/access.xml"/>

  <appender name="CONSOLE" class="ch.qos.logback.core.ConsoleAppender">
    <layout class="ch.qos.logback.classic.PatternLayout">
      <Pattern>
        %d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n
      </Pattern>
    </layout>
  </appender>

  <logger name="io.knotx" level="TRACE"/>
  <logger name="io.vertx" level="DEBUG"/>
  <logger name="com.codahale.metrics" level="TRACE"/>

  <root level="TRACE">
    <appender-ref ref="CONSOLE"/>
  </root>
</configuration>
``` 

## Operations configuration
Now we have to configure `knotx/conf/routes/operations.conf`. To get more familiar with it please check out [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/).

> In our config we should have define two operations:

Reviewer: `healthcheck` is already provided

- `healthcheck-operation`
- `example-cacheable-api-operation`

This operations are connected with `openapi.yaml`. As you remember, all endpoint definitions are connected with proper operations.
According to it let's fill `knotx/conf/routes/operations.conf` with following:
```hocon
routingOperations = ${routingOperations} [
  {
    operationId = example-cacheable-api-operation
    handlers = ${config.server.handlers.common.request} [
      {
        name = singleFragmentSupplier
        config = {
          type = json
          configuration {
            data-knotx-task = api-task
          }
        }
      },
      {
        name = fragmentsHandler
        config = { include required(classpath("routes/handlers/fragments-handler.conf")) }
      },
      {
        name = fragmentsAssembler
      }
    ] ${config.server.handlers.common.response}
  }
  {
    operationId = healthcheck-operation
    handlers = [
      {
        name = healthcheck
      }
    ]
  }
]
```

> Here we have defined all routing operations used in our tutorial, for us important is `example-cacheable-api-operation`.

Reviewer: ```suggestion
Here we have defined all routing operations used in our tutorial, for us, important is `example-cacheable-api-operation`.
```

First thing we should focus on is `handlers` array, there we can specify which handlers should be used to process our 
request. In our case, first handler to be called is `singleFragmentSupplier`. It's used to convert incoming http request
into `knotx-fragment` and assign `data-knotx-task` to it, task definition will be provided soon. Next to be called is
`fragmentsHandler`, here we have to attach additional configs with tasks specification. The last one is `fragmentsAssembler`
Purpose of using it, is to combine all fragments into final response.

## Handlers definition

> Last thing to do, is to define custom handler (Last thing about Knot.x configs, we have to create also Wiremock server). 

Reviewer: ```suggestion
Last thing to do, is to define custom handler. 
```

Author: fixed

> As I mentioned above it will be `fragments-handler`.

Reviewer: ```suggestion
As mentioned above it will be `fragments-handler`.
```

Author: fixed

> To do it let's remove from `knotx/conf/routes/handlers/` all handlers provided by default and create in this directory file

Reviewer: if you remove all configs, the `healthcheck` will stop working

Author: in this directory provided is only one defaut handler, it will not affect healthcheck if directory content will be removed

`fragments-handler.conf` with following structure:
```hocon
tasks {
  api-task {
    action = fetch-product-with-cache
      onTransitions {
        _success {
          action = product-to-body
        }
    }
  }
}

actions {
  fetch-product-with-cache {
    factory = in-memory-cache
    config {
      cache {
        maximumSize = 1000
        ttl = 5000
      }
      cacheKey = "product-{param.id}"
      payloadKey = fetch-product
    }
    doAction = fetch-product
  }

  product-to-body {
    factory = payload-to-body
    config {
      key = fetch-product
    }
  }

  fetch-product {
    factory = http
    config {
      endpointOptions {
        path = /product/id
        domain = webapi
        port = 8080
        allowedRequestHeaders = ["Content-Type"]
      }
    }
  }
}
```

> In the top of this file we have tasks definition, our config contains of only one task which is `api-task`. There we point some 

Reviewer: ```suggestion
At the top of this file we have tasks definition, our config consists of only one task which is `api-task`. There, we point some 
```

Author: fixed

When all of above will succeed `product-to-body` will be called and request will be finally properly assembled. To do it
we'll use `payload-to-body` factory, moreover we're pointing here `key = fetch-product`, as I mentioned above this depending 
of this value, proper payload will be injected here.

## Wiremock 
Wiremock is simply used for mocking server, we don't have to create real REST API, for our training purposes it's enough.
To do it let's create following directories on the project level directory:
- `common-services/webapi/__files/`
- `common-services/webapi/mappings/`

Into first directory please paste `product.json` file with following body:
```json
{
  "id": 21762532,
  "url": "http://knotx.io",
  "label": "Product"
}
```

And into second directory please paste `product.json` file with following body:
```json
{
  "request": {
    "method": "GET",
    "url": "/product/id"
  },
  "response": {
    "status": 200,
    "fixedDelayMilliseconds": 100,
    "bodyFileName": "product.json"
  }
}
```

Make sure that in `api-cache.yml` property `volume` properly points to `webapi` directory.

## Build and run
From project level directory execute following commands:
```shell script

> sh gradlew clean build

Reviewer: ```suggestion
gradlew clean build
```

Author: fixed

---
title: Caching HTTP requests
author: oskarjerzyk
keywords: tutorial
date: 2019-10-21
layout: tutorial
knotxVersions:
    - 2.0.0
---
# Overview

> In this tutorial we will add ability to cache http request to project based on [Knot.x Starter Kit](https://github.com/Knotx/knotx-starter-kit) 

Author: fixed

> template.

Author: fixed

> We will use existing `ActionFactory` from [knotx-fragments](https://github.com/Knotx/knotx-fragments) to achieve our goal

Reviewer 3: ```suggestion
The caching functionality can be implemented in a custom handler. Then we expose our custom handler in Open API specification. However, this approach requires development effort. But what if we would need to add a circuit breaker mechanism in future.

With [Knot.x Configurable Integrations](http://knotx.io/blog/configurable-integrations/) decoration of the target API  with cache or circuit breaker is a matter of configuration, not coding. 
```

> which will be implementing custom endpoint with cache ability.

Reviewer 3: To remove.

## Prerequisites

> - JDK 8

Reviewer 3: We've forgotten about Gradle, in other tutorials too. Can you please add it here and check other tutorials (a separate PR please).

Reviewer: is Gradle really needed if we use `gradlew`?

Author: I think it's not needed, gradlew will download appropriate gradle version and we don't care about it

- Docker

## What you're going to learn:

> - How to add cache ability to API

Reviewer 3: ```suggestion
- How to wrap an existing (target) API with caching proxy
```

> - How to set up Wiremock (we don't need to implement real API, mock API will be enough for our training purposes)

Reviewer 3: ```suggestion
- How to use [behaviours](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#behaviours) and [actions](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#actions)
```

Author: fixed

> - How to deploy project to Docker Swarm

Reviewer 3: We can remove this, it is not tutorial about Docker Swarm. However, we can mention about Docker Swarm in `Prerequisites`.

Author: fixed

Download the [Latest Knot.x Starter Kit release](https://github.com/Knotx/knotx-starter-kit/releases) and unzip it.

Project has the following structure:
```
├── docker
|   ├── Dockerfile                // Docker file with image definition.
├── functional                    // Keep here your functional tests. Example implementation included
├── gradle                        // Gradle wrapper and common gradle scripts
├── knotx/conf                    // Knotx configuration which will be copied to docker image
├── modules                       // Sub-modules of your project
│   ├── ...                       // example modules implementation
```

## Project configuration

> The first thing we need to do is to specify configuration files. Let's start in the project level directory and simply create `api-cache.yml`

Reviewer 3: What do you think about this oder:
- I use Configurable Integrations so I want to use tasks
- I configure HTTP action to invoke my target API
- then I configure in-memory cache and explain that it can be exchange with Redis very easily
- now specify task and and configure Open API / operation and Handlers
- build image, it is ready to use
- no we need to provider an environment, we choose Docker Swarm
- for target API we use Wiremock and its configuration looks like...
- let's start Docker Swarm and open in browser http://localhostL8092/..., refresh a few times
- end.

Author: fixed
